### PR TITLE
Al/815 add s3 adapter to store the payload

### PR DESCRIPTION
**Pull request by alouvion-aneo**
fix https://github.com/aneoconsulting/ArmoniK/issues/815
Epic : https://github.com/aneoconsulting/ArmoniK/issues/810
new S3 adapter is added. It can store payloads.
Docker factory updated.
Terraform scripts for newcoming factory in other PR : https://github.com/aneoconsulting/ArmoniK/issues/840

### CODE REVIEW

File: Adaptors/S3/src/Options/S3Storage.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+// 
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace ArmoniK.Core.Adapters.S3.Options;
+
+public class S3
```
Comment (aneojgurhem) [inline Adaptors/S3/src/Options/S3Storage.cs:27]: file name does not match class name

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -100,7 +100,7 @@ jobs:
 
     - name: Install ReSharper
       run: |
-        dotnet tool install -g JetBrains.ReSharper.GlobalTools --version 2022.2.2
+        dotnet tool install -g JetBrains.ReSharper.GlobalTools --version 2022.3.1
```
Comment (lemaitre-aneo) [inline .github/workflows/build.yml:103]: This should not change. If you want to update resharper in the pipeline, do it in a separate PR, but be aware that it can actually change the way code is formatted. That is why it has been hardcoded.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -148,6 +149,11 @@ jobs:
       with:
         dotnet-version: 6.x
 
+    - name: Minio Server UP
+      uses: isbang/compose-action@v1.4.0
+      with:
+        compose-file: "./docker-compose/S3/docker-compose-minio.yml"
```
Comment (lemaitre-aneo) [inline .github/workflows/build.yml:155]: Could you add Minio to the terraform deployment? We plan to remove docker compose from core in the (not so distant) future.

File: Adaptors/S3/tests/Usings.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY, without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+global using NUnit.Framework;
```
Comment (lemaitre-aneo) [inline Adaptors/S3/tests/Usings.cs:25]: I am not fond of `global using` and its non local effects. I believe it is the only place were it is used.
Here it is especially not needed as you have a single `.cs` file.

File: Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="AWSSDK.S3" Version="3.7.101.46" />
+    <PackageReference Include="CommunityToolkit.HighPerformance" Version="8.0.0" />
```
Comment (lemaitre-aneo) [inline Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj:11]: I believe you do not use it.

File: .github/workflows/build.yml (hunk 2 above)
Comment (alouvion-aneo) [reply .github/workflows/build.yml:103]: I agree, it was just a test to see if newer version was improving perfs, but not really

File: Adaptors/S3/tests/Usings.cs (hunk 4 above)
Comment (alouvion-aneo) [reply Adaptors/S3/tests/Usings.cs:25]: Same for me, it was generated by default when I have added the new project, so I was thinking maybe it's the new microsoft best practice

File: Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj (hunk 5 above)
Comment (alouvion-aneo) [reply Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj:11]: I still don't know, I'm testing it with my bench project.

File: Common/src/Storage/IObjectStorage.cs [hunk 6]
```diff
@@ -92,21 +92,6 @@ public static IObjectStorage CreateResourcesStorage(this IObjectStorageFactory f
...
-  ///   Update data if it already exists
-  /// </summary>
-  /// <param name="key">Key representing the object</param>
-  /// <param name="valueChunks">Chunks of data that will be stored in the Object Storage</param>
-  /// <param name="cancellationToken">Token used to cancel the execution of the method</param>
-  /// <returns>
-  ///   Task representing the asynchronous execution of the method
-  /// </returns>
-  /// <exception cref="ArmoniKException">there is 0 chunk</exception>
-  Task AddOrUpdateAsync(string                   key,
```
Comment (lemaitre-aneo) [inline Common/src/Storage/IObjectStorage.cs:106]: Why did you remove this method from the interface?

File: .github/workflows/build.yml (hunk 3 above)
Comment (alouvion-aneo) [reply .github/workflows/build.yml:155]: I will do it afterwards : https://github.com/aneoconsulting/ArmoniK/issues/840

File: Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj (hunk 5 above)
Comment (alouvion-aneo) [reply Adaptors/S3/src/ArmoniK.Core.Adapters.S3.csproj:11]: In the current implementation I finaly use it.
method comments are : Returns a <see cref="Stream"/> wrapping the contents of the given <see cref="ReadOnlyMemory{T}"/> of <see cref="byte"/> instance.
For the first implementation I think it is good to use it.
If we find better way to implement S3 read/write perfs in the bench part I will update it.

File: Common/src/Storage/IObjectStorage.cs (hunk 6 above)
Comment (alouvion-aneo) [reply Common/src/Storage/IObjectStorage.cs:106]: seen with you

File: Adaptors/S3/src/ObjectStorage.cs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+      logger_.LogError("The key {key} was not found.",
+                       key);
+      throw new ObjectDataNotFoundException("Key not found");
+    }
+
+    // Get the data from the response stream
+    using var reader      = new StreamReader(response.ResponseStream);
+    var       fileContent = reader.ReadToEnd();
+
+    var valuesCount = int.Parse(fileContent!);
```
Comment (aneojgurhem) [inline Adaptors/S3/src/ObjectStorage.cs:117]: You should test if it is null and throw a proper exception if it is the case instead of the suppressing the nullable with `!`

File: Adaptors/S3/src/ObjectStorage.cs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                                               .ConfigureAwait(false);
+    return deleteObjectsResponse.DeletedObjects.Count == valuesCount + 1;
+  }
+
+  /// <inheritdoc />
+  public IAsyncEnumerable<string> ListKeysAsync(CancellationToken cancellationToken = default)
+    => throw new NotImplementedException();
+}
+
+internal static class S3StorageHelper
```
Comment (aneojgurhem) [inline Adaptors/S3/src/ObjectStorage.cs:190]: This is an extension class for `AmazonS3Client ` so it should be called `AmazonS3ClientExt`

**Review by aneojgurhem — APPROVED** [review]
